src/main/java/io/cryostat/discovery/KubeApiDiscovery.java

```diff
@@ -0,0 +1,623 @@
+
+                ownerNode.children.add(childNode);
+                childNode.parent = ownerNode;
+
+                ownerNode.persist();
+                childNode.persist();
+
+                child = owner;
+            }
+
+            nsNode.children.add(child.getRight());
+            child.getRight().parent = nsNode;
+        } else {
+            // if the Endpoint points to something else(?) than a Pod, just add the target straight
+            // to the Namespace
+            nsNode.children.add(targetNode);
+            targetNode.parent = nsNode;
+            targetNode.persist();
+        }
+
+        nsNode.persist();
+    }
+
+    private Pair<HasMetadata, DiscoveryNode> getOwnerNode(Pair<HasMetadata, DiscoveryNode> child) {
+        HasMetadata childRef = child.getLeft();
+        if (childRef == null) {
+            return null;
+        }
+        List<OwnerReference> owners = childRef.getMetadata().getOwnerReferences();
+        // Take first "expected" owner Kind from NodeTypes, or if none, simply use the first owner.
+        // If there are no owners then return null to signify this and break the chain
+        if (owners.isEmpty()) {
+            return null;
+        }
+        String namespace = childRef.getMetadata().getNamespace();
+        OwnerReference owner =
+                owners.stream()
+                        .filter(o -> KubeDiscoveryNodeType.fromKubernetesKind(o.getKind()) != null)
+                        .findFirst()
+                        .orElse(owners.get(0));
+        return discoveryNodeCache.computeIfAbsent(cacheKey(namespace, owner), this::queryForNode);
+    }
+
+    private Triple<String, String, String> cacheKey(String ns, OwnerReference resource) {
+        return Triple.of(ns, resource.getKind(), resource.getName());
+    }
+
+    // Unfortunately, ObjectReference and OwnerReference both independently implement getKind and
+    // getName - they don't come from a common base class.
+    private Triple<String, String, String> cacheKey(String ns, ObjectReference resource) {
+        return Triple.of(ns, resource.getKind(), resource.getName());
+    }
+
+    private Pair<HasMetadata, DiscoveryNode> queryForNode(
+            Triple<String, String, String> lookupKey) {
+
+        String namespace = lookupKey.getLeft();
+        String name = lookupKey.getRight();
+        KubeDiscoveryNodeType nodeType =
+                KubeDiscoveryNodeType.fromKubernetesKind(lookupKey.getMiddle());
+        if (nodeType == null) {
+            return null;
+        }
+
+        synchronized (queryLocks.computeIfAbsent(lookupKey, k -> new Object())) {
+            HasMetadata kubeObj =
+                    nodeType.getQueryFunction().apply(client()).apply(namespace).apply(name);
+
+            DiscoveryNode node =
+                    DiscoveryNode.getNode(
+                                    n -> {
+                                        return name.equals(n.name)
```

Comment:
So this label is used for the case where Namespace A and Namespace B both have Endpoints objects named E, is that right? This way we can unambiguously use the `<name, namespace>` pair as an identifier for that particular node.

Reply:
Oh yes it is indeed for identify nodes of the same names in different namespaces. I used it mostly for identifying intermediate parent nodes tho (i.e. pod, replicaset, deployment, etc.).

> This way we can unambiguously use the <name, namespace> pair as an identifier for that particular node.

That means we should return `Pair<HasMetadata, <Pair<String, DiscoveryNode>>` from this method `queryForNode`?

Reply:
Oh no, you don't need to make any changes here, I was just trying to make sure my reading of it was correct.

Reply:
Tho, that being said, that should also check for nodeType in case there might be 2 nodes of the same name, in the same namespace but different type.